### DISCUSSION TITLE
Trying to run generate from grammar with a node container

### DISCUSSION THREAD

**Discussion by jonnyso** [nvim-treesitter/nvim-treesitter#3926] [Q&A] [answered]
Hey there,

I'm trying to execute TSUpdate but I don't have node installed locally, I've created a simple shellscript for running node and put it in my PATH like below:

`podman run --rm -i -w /usr/bin/app -v "./:/usr/bin/app:z" node node $@`

When I run checkhealth everything seems to pick up node correctly, however when I try TSUpdate it gets stuck on this part:

[nvim-treesitter] [0/1] Generating source files from grammar.js...

I can't figure out what I'm missing here, I already confirmed that giving it a tty will cause other errors.

**Comment by theHamsta** [comment, marked as answer]
Which node container are you using? `node:latest`?

**Reply by jonnyso** [reply]
Yes

**Reply by theHamsta** [reply]
I'm getting
```
nvim-treesitter[c]: Error during "tree-sitter generate"                                                                                                                                                                                                                                                                                                           
Error: OCI runtime error: runc create failed: unable to start container process: can't get final child's PID from pipe: EOF
Node process exited with status 126
Press ENTER or type command to continue
```
I have no experience with Podman

**Reply by theHamsta** [reply]
let me check my docker/container setup on this machine....

**Reply by theHamsta** [reply]
I also fails outside of nvim. I try to fix it first...

**Reply by theHamsta** [reply]
When tree-sitter-cli invokes the node binary, it's important that `stdin` and `stdout` get correctly forwarded from your bash script to the container process. Is this happening right now? You probably need to do something with `&0`/`&1`/`&2`. tree-sitter-cli will throw JS code via stdin to node and expects a JSON from stdout.

dsl.js is loaded here
https://github.com/tree-sitter/tree-sitter/blob/870fb8772f3e47b5aef4c7000a78d61f3aa2b005/cli/src/generate/mod.rs#L195-L171

**Reply by jonnyso** [reply]
Thanks ! That makes sense ! I'll see what I can do here.

**Reply by theHamsta** [reply]
The node process will also need to have access to the grammar's path https://github.com/tree-sitter/tree-sitter/blob/870fb8772f3e47b5aef4c7000a78d61f3aa2b005/cli/src/generate/dsl.js#L417. It must be the same path in the container as outside of the image. The environment variable is coming from here https://github.com/tree-sitter/tree-sitter/blob/870fb8772f3e47b5aef4c7000a78d61f3aa2b005/cli/src/generate/mod.rs#L172

**Reply by jonnyso** [reply]
Very helpful, thanks !

**Reply by theHamsta** [reply]
It might be easier for debugging to adapt this code here https://github.com/tree-sitter/tree-sitter/blob/870fb8772f3e47b5aef4c7000a78d61f3aa2b005/cli/src/generate/mod.rs#L170-L190 to launch podman with the correct arguments. You can build the repo by `cargo install --path ./cli` in the tree-sitter repo

**Reply by theHamsta** [reply]
I personally use https://github.com/nvm-sh/nvm to avoid using node from my system package manager

**Reply by jonnyso** [reply]
That's actually much better ahaha I'll go with that
